restructure calculation of newdq

gsavva opened the pull request:
use ZGEMM (instead if explicit loops) to calculate the contributions to D-matrix for all atoms of a given type in one-shot.

This means calling ZGEMM as many times as the atom *types*.
Two alternatives have been implemented [tmp = potential * CONJG(phases)]:
- res1 = (tmp)^T * CONJG(qpw)
- res2 = (qpw)^H * tmp

TODO:
- [x] test performance

gsavva committed:
add comments

gsavva committed:
use ZGEMM (instead if explicit loops) to calculate the contributions to D-matrix for all atoms of a given type in one-shot.

This means calling ZGEMM as many times as the atom *types*.
For now, implementation is alongside the native QE implementation to check correctness.
Two alternatives have been implemented [tmp = potential * CONJG(phases)]:
- res1 = (tmp)^T * CONJG(qpw)
- res2 = (qpw)^H * tmp

gsavva committed:
remove diff

gsavva committed:
comment-out QE implementation

gsavva committed:
use res1 = (tmp)^T * CONJG(qpw)

gsavva committed:
use res2 = (qpw)^H * tmp

gsavva committed:
remove commented code

gsavva commented:
Fails for systems with 2 spins in `Release` compilation.
In `Debug` compilation, it _seems_ to run fine... :confused:

toxa81 commented on LR_Modules/newdq.f90:
```diff
@@ -78,53 +80,72 @@ subroutine newdq (dvscf, npe)
+               !=============== compute Q*V for all atoms of type nt
+               !    DGEMM( TA,  TB,    M,   N,   K, ALPHA,   A, LDA,                        B, LDB,  BETA,   C,  LDC)
+!              call ZGEMM('T', 'N', N_nt, nij, ngm, 1.0d0, tmp, ngm, CONJG(atom_type(nt)%qpw), ngm, 0.0d0, res1, N_nt)
+               call ZGEMM('C', 'N', nij, N_nt, ngm, 1.0d0, atom_type(nt)%qpw, ngm, tmp, ngm, 0.0d0, res2, nij)
```
for zgemm you probaly need to pass dcmplx(1.d0, 0.d0) and dcmplx(0.d0, 0.d0) - complex alpha and beta numbers

gsavva committed:
fix in calling ZGEMM

use complex "dcmplx(1.d0, 0.d0)" for ALPHA and "dcmplx(0.d0, 0.d0)" for BETA, and not just double

gsavva commented:
Two alternatives have been implemented [tmp = potential * CONJG(phases)]:
1. res1 = (tmp)^T * CONJG(qpw)
2. res2 = (qpw)^H * tmp

Regarding performance:
- the 1st one, `res1 = (tmp)^T * CONJG(qpw)`, is slower than the native QE implementation
- the 2nd one, `res2 = (qpw)^H * tmp`, needs ~ 1/5th of the time needed by the 1st.

==> keeping `res2 = (qpw)^H * tmp`

toxa81 commented:
@gsavva `lr_addusdens` branch contains two allocations of qpw with swapped dimensions. One of them is temporary. I will check which one of two is needed, clean it up and then we can merge to the main branch.

gsavva committed:
clean-up

gsavva committed:
add OpenMP in newdq

gsavva committed:
clean-up + add clocks

gsavva committed:
clean-up of unnecessary variables

toxa81 committed:
rename qpw1 to qpw_t and keep it

toxa81 commented:
@gsavva I decided to keep both qpw and qpw-transposed at the moment. Once the PR is built I think it is good to be merged

toxa81 commented:
cscs-ci run default

toxa81 reviewed (approved):
LGTM
